Feature | Postgres Support

ksimenic opened the pull request:
This PR merges 2 existing PRs: #60 and #63 and potentially solves https://github.com/Sammyjo20/laravel-haystack/issues/61

Additionally, both PRs are standardised (pint) and upgraded with following changes:
- 60 - improved description in config file
- 60 - improved env vars fallbacks and added new env var as override
- 60 - added missing return types on models
- 60 - improve TestCase logic, add return types and fix failing test
- 63 - extracted `DB::connection instanceof` check to separate function and take db_connection configuration var (from 60) into account

ksimenic commented:
Additional comment for testing. Since sqlite is used for testing, it is not that straight forward to implement tests for mysql/pgsql. That would also require additional work for both local development and github actions. However, if adding tests is necessary I can work on them in the scope of this PR or in some other PR.

Sammyjo20 commented:
Many thanks for this PR I will have a review now. If you know how to build a testing matrix with SQLite, MySQL and Postgres that would be fantastic. Might that be something you could add to this PR please?

Sammyjo20 commented on src/Helpers/SerializationHelper.php:
```diff
@@ -0,0 +1,46 @@
+    {
+        $serialized = serialize($value);
+
+        return self::isPgsql()
+            ? base64_encode($serialized)
+            : $serialized;
```
This doesn't need to be put on two lines

```suggestion
        return self::isPgsql() ? base64_encode($serialized) : $serialized;
```

ksimenic replied:
Fixed.

Sammyjo20 commented on src/Helpers/SerializationHelper.php:
```diff
@@ -0,0 +1,46 @@
+        string $serialized,
+        array $options = []
+    ): mixed {
+        if (
+            self::isPgsql()
+            && ! Str::contains($serialized, [':', ';'])
+        ) {
```
```suggestion
        if (self::isPgsql() && ! Str::contains($serialized, [':', ';'])) {
```

ksimenic replied:
Fixed.

Sammyjo20 commented on src/Helpers/SerializationHelper.php:
```diff
@@ -0,0 +1,46 @@
+    /**
+     * Unserialize the given value.
+     */
+    public static function unserialize(
+        string $serialized,
+        array $options = []
+    ): mixed {
```
```suggestion
    public static function unserialize(string $serialized, array $options = []): mixed 
    {
```

ksimenic replied:
Fixed.

Sammyjo20 reviewed (changes requested):
Thank you for this

ksimenic commented:
@Sammyjo20 

Added tests, as requested. Additionally, I ran pint on whole project (not just on updated files) and fixed code style per your suggestion.

There are 2 things you should be aware of.

1. Running tests locally
Since now we have tests on MySQL and PgSQL, to run the tests locally we need both MySQL and PgSQL server. 
I am using DBnging locally. Both MySQL and PgSQL are created with default values except the port. For MySQL port is 33066 and for PgSQL port is 54321. Additionally, once server is created and started you must create database named "testing" on both servers.
<img width="525" alt="image" src="https://user-images.githubusercontent.com/12883124/220610094-776d1046-82f2-44d8-a095-9ff10adebfca.png">

2. Running tests in matrix
Needed to drop support for "windows-latest" since you can't create "services" (postgres, mysql) there.
<img width="456" alt="image" src="https://user-images.githubusercontent.com/12883124/220611095-243f58f8-561d-4bb2-930a-e50a47d3c24d.png">

ksimenic commented:
Hey @Sammyjo20. I am just wondering if there is a plan to release this any time soon? If not, I will need to unblock my projects by using my own fork. No problem at all if there is no plan to release it. Thank you!

Sammyjo20 commented:
Hey @ksimenic I am hoping to release this at the weekend if this works for you?

ksimenic commented:
@Sammyjo20 works for me. Thank you!

Sammyjo20 reviewed (approved):
Happy with this 💪

Sammyjo20 commented:
Thank you @ksimenic for your help on this!

ksimenic commented:
Hey @Sammyjo20. Thank you for releasing this! Just wondering why you dropped support for Laravel 8.x?

Sammyjo20 commented:
Hey @ksimenic I've dropped support for Laravel 8 because it's no longer getting any security updates or officially supported. 

![image](https://user-images.githubusercontent.com/29132017/222842001-5b54b9a4-b043-4a89-a042-636acb451ca1.jpeg)
